pkg/engine/diff.go

```diff
@@ -230,7 +230,19 @@ func printObject(
 // GetResourceOutputsPropertiesString prints only those properties that either differ from the input properties or, if
 // there is an old snapshot of the resource, differ from the prior old snapshot's output properties.
 func GetResourceOutputsPropertiesString(
-	step StepEventMetadata, indent int, planning bool, debug bool) string {
+	step StepEventMetadata, indent int, planning bool, debug bool, refresh bool) string {
+	// We should only print outputs if the outputs are known to be complete. This will be the case if we are
+	//   1) not doing a preview
+	//   2) doing a refresh
+	//   3) doing a read
+	//
+	// Technically, 2 and 3 are the same, since they're both bottoming out at a provider's implementation of Read, but
+	// the upshot is that either way we're ending up with outputs that are exactly accurate. If we are not sure that we
+	// are in one of the above states, we shouldn't try to print outputs.
+	if planning && (!refresh || step.Op == deploy.OpRead || step.Op == deploy.OpReadReplacement) {
+		return ""
+	}
```

Comment:
Nit: my brain sucks at pulling apart boolean expressions like this.  i would honestly be fine with separate if statements, each explaining why they work the way we do.  i.e. you can just say:

```go
if planning {
    // during preview we don't normally care to print out outputs.  However, there are a few cases were we still want to.
    if (!refresh) {
        // during a refresh, we still want to print out output even during a preview, because yadda yadda.  for anythin
    }

   // it's a refresh during preview.  but it's for a read/replace.  we don't want to show that because reasons... yadda yadda
}
```

Reply:
will do.